lib/__init__.py

```diff
@@ -112,15 +118,81 @@ def connect(dsn=None, connection_factory=None, cursor_factory=None, **kwargs):
     library: the list of supported parameters depends on the library version.
 
     """
+    
     kwasync = {}
     if 'async' in kwargs:
         kwasync['async'] = kwargs.pop('async')
     if 'async_' in kwargs:
         kwasync['async_'] = kwargs.pop('async_')
 
+    lbprops = LoadBalanceProperties(dsn, **kwargs)
+    if lbprops.hasLoadBalance() :
+        conn = getConnectionBalanced(lbprops, connection_factory, cursor_factory, **kwasync)
+        if conn != None:
+            return conn
+        print('Failed to apply load balancing, Trying normal connection')
+    
+    dsn = lbprops.getStrippedDSN()
+    kwargs = lbprops.getStrippedProperties()
     dsn = _ext.make_dsn(dsn, **kwargs)
     conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
     if cursor_factory is not None:
         conn.cursor_factory = cursor_factory
-
     return conn
+
+def getConnectionBalanced(lbprops, connection_factory, cursor_factory=None, **kwasync):
+    kwargs = lbprops.getStrippedProperties()
+    loadbalancer = lbprops.getAppropriateLoadBalancer()
+    dsn = lbprops.getStrippedDSN()
+    unreachableHosts = loadbalancer.getUnreachableHosts()
+    failedHosts = unreachableHosts
+    chosenHost = loadbalancer.getLeastLoadedServer(failedHosts)
+    dsn = _ext.make_dsn(dsn,**kwargs)
+    needsRefresh = loadbalancer.needsRefresh()
+    if chosenHost == '' or needsRefresh:
+        controlConnection = _connect(dsn, connection_factory=connection_factory, **kwasync)
+        if cursor_factory is not None:
+            controlConnection.cursor_factory = cursor_factory
+        if not loadbalancer.refresh(controlConnection):
+            return None
+        if needsRefresh:
+            dsnhost = controlConnection.info.host_addr
+            loadbalancer.updateConnectionMap(dsnhost, 1)
+            loadbalancer.updateConnectionMap(chosenHost, -1)
```

Comment:
Yes chosenhost is not always non-empty here but when decreasing the chosenHost count, if the number of connections is 0 it will not decrease it further. I have decreased the chosenHost count by 1 because when new control connection is made for refresh, it makes it using the original dsn but the host is chosen in the previous lines and that increases the connection count for that host. So since no connection is being made to that chosenHost, I decrease the connection count here.